File: Python/src/dss/collections/linkedlist_single.py

import colorama


class SllNode:
  def __init__(self, _val=0, _next=None) -> None:
    self.value = _val
    self.next = _next


class SinglyLinkedList:
  def __init__(self):
    self.head = None
    self.tail = None
    self.__size = 0

  def __len__(self):
    return self.__size

  def __str__(self):
    curr = self.head
    node_values = []
    while curr:
      node_values.append(colorama.Fore.GREEN + str(curr.value))
      node_values.append(colorama.Style.RESET_ALL + "->")
      curr = curr.next
    return "".join(node_values[:len(node_values)-1])

  ### Access ###
  def get(self, index):
    if index < 0 or index >= self.__size:
      raise IndexError("Index out of range")

    curr = self.head
    while index > 0:
      curr = curr.next
    return curr

  ### Insertions ###
  def append(self, value):
    node = SllNode(value)
    if not self.head:
      # no elements in list
      self.tail = node
      self.head = self.tail
    else:
      self.tail.next = node
      self.tail = self.tail.next

    self.__size += 1
    return True

  def append_front(self, value):
    node = SllNode(value)
    if not self.head:
      # no elements in list
      self.head = node
      self.tail = self.head
    else:
      node.next = self.head
      self.head = node

    self.__size += 1
    return True

  def insert(self, index, value):
    if index == self.__size:
      self.append(value)
    elif index == 0:
      self.append_front(value)
    else:
      prev = self.get(index - 1)
      node = SllNode(value, prev.next)

      prev.next = node

      self.__size += 1
    return True

    ### Deletions ###

  ### Deletions ###
  def pop(self):
    if not self.head:
      return None

    node = self.tail
    if self.__size == 1:
      self.head = self.tail = None
    else:
      self.tail = self.get(self.__size - 2)
      self.tail.next = None

    self.__size -= 1
    return node

  def pop_front(self):
    if not self.head:
      return None

    node = self.head
    if self.__size == 1:
      self.head = self.tail = None
    else:
      self.head = self.head.next

    self.__size -= 1

    return node

  def remove(self, index):
    if index == self.__size:
      return self.pop()
    if index == 0:
      return self.pop_front()

    prev = self.get(index - 1)

    node = prev.next
    prev.next = prev.next.next

    self.__size -= 1
    return node

  ### Utils ###

  ### Private Functions ###

File: NodeJS/src/collections/linked-list-single.js
export { SinglyLinkedList };

class SlNode {
  constructor(_value = 0, _next = null) {
    this.value = _value;
    this.next = _next;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /** Access */
  get(index) {
    //O(n)
    if (index < 0 || index >= this.length) return undefined;

    let curr = this.head;
    for (let i = 0; i < index; ++i) {
      curr = curr.next;
    }

    return curr;
  }

  /** Updates */
  set(index, value) {
    let node = this.get(index);
    if (!node) return false;

    node.value = value;

    return true;
  }

  /** Insertions */
  prepend(value) {
    //O(1)
    let node = new SlNode(value, this.head);

    if (!this.head) this.tail = node;
    this.head = node;

    ++this.length;
  }

  append(value) {
    //O(1)
    let node = new SlNode(value);

    if (!this.head) {
      // Edge Case: We currently have no elements in the list.
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = this.tail.next;
    }

    ++this.length;
  }

  insert(index, value) {
    if (index < 0 || index > this.length) return false;

    if (index === this.length) this.append(value);
    else if (index === 0) this.prepend(value);
    else {
      let pre = this.get(index - 1);
      let node = new SlNode(value, pre.next);

      pre.next = node;

      ++this.length;
    }
    return true;
  }

  /** Deletions */
  popFront() {
    //O(1)
    if (!this.head) return undefined;

    let node = this.head;
    this.head = this.head.next;

    --this.length;

    //Edge Case: emptied list
    if (!this.head) this.tail = null;

    return node;
  }

  pop() {
    //O(n)
    if (!this.head) return undefined;

    let node = this.head;

    if (this.length > 1) {
      while (node.next !== this.tail) node = node.next;

      // swap node and tail;
      this.tail = node;
      node = node.next;

      // update tail.next
      this.tail.next = null;
    } else {
      //Edge Case: 1 item in list
      this.head = null;
      this.tail = null;
    }

    --this.length;

    return node;
  }

  remove(index) {
    if (index < 0 || index >= this.length) return false;

    if (index === 0) this.popFront();
    else if (index === this.length - 1) this.pop();
    else {
      let pre = this.get(index - 1);
      pre.next = pre.next.next;

      --this.length;
    }

    return true;
  }

  /** Utility */
  isEmpty() {
    return this.head == null;
  }

  /** Iterators */
  [Symbol.iterator]() {
    let node = this.head;
    return {
      next: () => {
        if (!node) return { done: true };

        const value = node.value;
        node = node.next;
        return {
          done: false,
          value,
        };
      },
    };
  }
}
